### PR TITLE
[Event Hubs Client] Track One: Ephemeral Test Resources

### PR DESCRIPTION
**Pull request by jsquire**
# Summary

The intent of these changes is to revise the approach used for dynamic resource management by the test suite, shifting to an approach where an ephemeral Event Hubs namespace and Azure Storage account are created upon request by a Live test and remains only for the duration of a given test run.

### Goals

- Reduce contention on a single Event Hubs namespace when tests are run; full parallelization of the track one and track two tests suites for each of their associated platforms  has begun to overwhelm the management plane with requests against a single namespace.  

- Reduce Azure spend for the cost center by limiting the time that charges are accrued by having an active resource only when being used.   This also allows for a greater number of throughput units to be allocated without increasing spend.

- Tune retry configuration to account for additional contention on Azure resources due to parallelization.

- Revise the associated assets and documentation to align with the changes.

# Last Upstream Rebase

Thursday, August 15, 2019  1:34pm (EDT)

### CODE REVIEW

File: sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1 [hunk 1]
```diff
@@ -0,0 +1,383 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (conniey) [inline sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1]: Is there a unified way to do this across different client libraries running live services? I remember reading a thread in Teams, but curious why Powershell over ARM templates?

**Review by conniey — APPROVED** [review]

File: sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1 (hunk 1 above)
Comment (jsquire) [reply sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1]: Service Principals and permissions are not, as far as I know or was able to determine, available in ARM.... so, Powershell.  

There's been a little bit of chatter about centralizing, but I don't think the approach or concept is fully baked yet and was mostly centered around the more standard needs of services using recorded tests with the HTTP pipeline, at present.

File: sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1 [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+  $indent = "    "
+
+  Write-Host "`n"
+  Write-Host "Event Hubs Live Test Environment Setup"
+  Write-Host ""
+  Write-Host "$($indent)This script handles creation and configuration of needed resources within an Azure subscription"
+  Write-Host "$($indent)for use with the Event Hubs client library Live test suite."
+  Write-Host ""
+  Write-Host "$($indent)Upon completion, the script will output a set of environment variables with sensitive information which"
+  Write-Host "$($indent)fare used for testing.  When running Live tests, please be sure to have these environment variables available,"
```
Comment (kinelski) [inline sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1]: Typo: fare => are

File: sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1 [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+  Write-Host "$($indent)-SubscriptionName`t`tRequired.  The name of the Azure subscription to be used for"
+  Write-Host "$($indent)`t`t`t`trunning the Live tests."
+  Write-Host ""
+    
+  Write-Host "$($indent)-ResourceGroupName`t`tThe name of the Azure Resource Group that will contain the resources"
+  Write-Host "$($indent)`t`t`t`tused for the tests.  This will be created if it does not exist."
+  Write-Host ""
+
+  Write-Host "$($indent)-ServicePrincipalName`tThe name to use for the service principal that will be"
+  Write-Host "$($indent)`t`t`t`tbe created to manage the Event Hub instances dynamically for the tests.  This"
```
Comment (kinelski) [inline sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1]: nit: repeated word. "that will be **be** created".

File: sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1 [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+
+    $scrambled = ($password.ToCharArray()) | Get-Random -Count ($password.Length)
+    return (-join $scrambled)
+}
+
+function TearDownResources
+{
+    <#
+      .SYNOPSIS
+        Cleans up any Azure resoruces created by the script.
```
Comment (kinelski) [inline sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1]: Typo: resources.

File: sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1 [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+        
+      .DESCRIPTION
+        Responsible for cleaning up any Azure resoruces created
```
Comment (kinelski) [inline sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1]: Typo: resources.

File: sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1 [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+  DisplayHelp $Defaults
+  exit 0
+}
+
+if ([String]::IsNullOrEmpty($AzureRegion))
+{
+    $AzureRegion = "southcentralus"
+}
+
+# Disallow prinicpal names with a space.
```
Comment (kinelski) [inline sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1]: Typo: principal.

File: sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1 [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+$resourceGroup = (Get-AzResourceGroup -ResourceGroupName "$($ResourceGroupName)" -ErrorAction SilentlyContinue)
+
+if ($resourceGroup -eq $null)
+{
+    $createResourceGroup = $true
+}
+
+if ($createResourceGroup)
+{
+    Write-Host "`t...Creaating new resource group"
```
Comment (kinelski) [inline sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1]: Typo: Creating.

File: sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1 [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+    $resourceGroup = (New-AzResourceGroup -Name "$($ResourceGroupName)" -Location "$($AzureRegion)")
+}
+
+if ($resourceGroup -eq $null)
+{
+    Write-Error "Unable to locate or create the resource goup: $($ResourceGroupName)"
```
Comment (kinelski) [inline sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1]: Type: group.

Comment (kinelski) [reply sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1]: I mean, "Typo". But I'll leave it here for the sake of irony.

File: sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1 [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+    exit -1
+}
+
+# At this point, we may have created a resource, so be safe and allow for removing any
+# resoruces created should the script fail.
```
Comment (kinelski) [inline sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1]: Typo: resources.

File: sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1 [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+
+try 
+{
+    # Create the service prinicpal and grant contributor access for management in the resource group.
```
Comment (kinelski) [inline sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1]: Typo: principal.

File: sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1 [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+        New-AzRoleAssignment -ApplicationId "$($principal.ApplicationId)" -RoleDefinitionName "Contributor" -ResourceGroupName "$($ResourceGroupName)" | Out-Null
+    }
+    catch 
+    {
+        Write-Host "`t...Still waiting for identity propagation (this will take a moment)"
+        Start-Sleep 60
+        New-AzRoleAssignment -ApplicationId "$($principal.ApplicationId)" -RoleDefinitionName "Contributor" -ResourceGroupName "$($ResourceGroupName)" | Out-Null
+    }    
+
+    # Write the environment variables
```
Comment (kinelski) [inline sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1]: Super nit: maybe we ending the comment with a period to keep it consistent with other comments in the file?

**Review by kinelski — APPROVED** [review]
I'm not a PowerShell expert, but it LGTM. Added a few comments regarding typos and small stuff.

**Comment by jsquire** [comment]
Obvious that the spelling checker does not follow me to where I wrote the PowerShell, eh?   :)

File: sdk/eventhub/Azure.Messaging.EventHubs/assets/live-tests-azure-setup.ps1 [hunk 12]
```diff
@@ -254,22 +254,22 @@ if ($resourceGroup -eq $null)
...
 {
-    Write-Host "`t...Creaating new resource group"
+    Write-Host "`t...Creating new resource group"
     $resourceGroup = (New-AzResourceGroup -Name "$($ResourceGroupName)" -Location "$($AzureRegion)")
 }
 
 if ($resourceGroup -eq $null)
 {
-    Write-Error "Unable to locate or create the resource goup: $($ResourceGroupName)"
+    Write-Error "Unable to locate or create the resource group: $($ResourceGroupName)"
```
Comment (conniey) [inline sdk/eventhub/Azure.Messaging.EventHubs/assets/live-tests-azure-setup.ps1]: I was hoping you'd edit it to `gout`

File: sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1 [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+  [ValidateScript({ $_.Length -ge 6})]
+  [string] $ServicePrincipalName,
+
+  [Parameter(Mandatory=$true, ParameterSetName="Execute")]
+  [AllowNull()]  
+  $AzureRegion = $null
+)
+
+# =====================
+# == Module Imports  ==
```
Comment (conniey) [inline sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1]:
```suggestion
# == Module Imports ==
```

File: sdk/eventhub/Microsoft.Azure.EventHubs/tests/Amqp/AmqpMEssageCoverterTests.cs [hunk 14]
```diff
@@ -55,7 +55,8 @@ private static ArraySegment<byte>[] ReadMessagePayLoad(AmqpMessage message, int
                 ArraySegment<byte>[] messageBuffers = message.GetPayload(payloadSize, out more);
                 if (messageBuffers != null)
                 {
-                    foreach (var segment in messageBuffers) { message.CompletePayload(segment.Count); }
+                    foreach (var segment in messageBuffers)
+                    { message.CompletePayload(segment.Count); }
```
Comment (conniey) [inline sdk/eventhub/Microsoft.Azure.EventHubs/tests/Amqp/AmqpMEssageCoverterTests.cs]: Would it be more .NET to put it on separate lines?
```cs
{
    message.CompletePayload(segment.Count);
}
```

Comment (jsquire) [reply sdk/eventhub/Microsoft.Azure.EventHubs/tests/Amqp/AmqpMEssageCoverterTests.cs]: Without a doubt;  however, this is track one code, so I'm making just the minimal set of changes.  :)

**Pull request merged**